### PR TITLE
Refactor resample algorithm to use pure functions

### PR DESCRIPTION
**Pull request by andyphelps**
* Algorithm now uses a snap to epoch approach allowing a sliding window approach to resampling between pings

### CODE REVIEW

File: spark/libraries/ais-schemata/src/main/scala/uk/gov/ukho/ais/SchemaTuple.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package uk.gov.ukho.ais
+
+import java.sql.Timestamp
+
+object SchemaTuple {
+
+  type PartitionedAisSchema = (String,
```
Comment (JoelClemence) [inline spark/libraries/ais-schemata/src/main/scala/uk/gov/ukho/ais/SchemaTuple.scala]: Can this include 'tuple' in the name to declash from the Schema in Schemata?

File: spark/resample-ais/src/main/scala/uk/gov/ukho/ais/resample/ResamplingFilter.scala [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      if (shouldInterpolateBetweenPings(prevPing, currPing)) {
+        addInterpolatedPings(prevPing, currPing)
+      } else {
+        Seq(currPing)
+      }
+    }
+
+    private def shouldInterpolateBetweenPings(prevPing: Ping, currPing: Ping)(
+        implicit config: Config): Boolean = {
+      currPing.mmsi.equals(prevPing.mmsi) &&
```
Comment (JoelClemence) [inline spark/resample-ais/src/main/scala/uk/gov/ukho/ais/resample/ResamplingFilter.scala]: Will mmsis ever be different?

File: spark/resample-ais/src/main/scala/uk/gov/ukho/ais/resample/ResamplingFilter.scala [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+      import df.sparkSession.implicits._
+      import org.apache.spark.sql.functions._
+      import ResampleAisPings.{prev, interpolateBetweenPingsFromTime}
+
+      val windowSpec = Window
+        .partitionBy(MMSI)
+        .orderBy(ACQUISITION_TIME)
+
+      df.withColumn(prev(ARKEVISTA_POS_ID),
```
Comment (JoelClemence) [inline spark/resample-ais/src/main/scala/uk/gov/ukho/ais/resample/ResamplingFilter.scala]: What does this do?

File: spark/resample-ais/src/test/scala/ComponentTest.scala [hunk 4]
```diff
@@ -124,20 +124,17 @@ class ComponentTest {
       "180207074437632\t3456793\t2017-01-01 00:18:00\t1.0035087719298246\t2.0035087719298246\tA\t1\t0\t0\t0.1\t352.2\t320\t-1\t0\t59916\t1024\twibble"
     )
 
-    assertThat(getLinesFromFiles("/year=2019/month=1/day=1"))
-      .as("Contains correct 2019 data")
```
Comment (JoelClemence) [inline spark/resample-ais/src/test/scala/ComponentTest.scala]: Please reinstate these, it makes debugging failed tests easier.

File: spark/resample-ais/src/test/scala/ResamplerTest.scala [hunk 5]
```diff
@@ -29,66 +31,61 @@ class ResamplerTest {
...
 
-  Session.init("ResamplerTestSession", isTestSession = true)
-
-  private val doubleComparator: Comparator[Double] = new Comparator[Double] {
-    override def compare(a: Double, b: Double): Int =
-      Precision.compareTo(a, b, DOUBLE_COMPARISON_PRECISION)
-  }
+  Session.init("", isTestSession = true)
 
   @Test
```
Comment (JoelClemence) [inline spark/resample-ais/src/test/scala/ResamplerTest.scala]: Can you add a test where the mmsis aren't the same?

File: spark/resample-ais/src/test/scala/ResamplerTest.scala [hunk 6]
```diff
@@ -97,115 +94,125 @@ class ResamplerTest {
...
       Seq((0, 0, 0), (3, 0.03, 0.03), (6, 0.06, 0.06)))
 
-    val outPings: Seq[ShipPing] = inputRdd.resample.collect()
+    val outPings: DataFrame = inputDataFrame.resample(TEST_CONFIG)
 
     assertThatOutPingsMatchesExpectedPings(expectedPings, outPings)
   }
 
   @Test
   def whenResamplingPointsOver30kmThenNoResamplingPerformed(): Unit = {
```
Comment (oliverCausley) [inline spark/resample-ais/src/test/scala/ResamplerTest.scala]: Does this need an additional test for the other parameter (100km)?

File: spark/resample-ais/src/test/scala/ResamplerTest.scala [hunk 7]
```diff
@@ -97,115 +94,125 @@ class ResamplerTest {
...
           (9996, 0.1, 0.1)))
 
-    val outPings: Seq[ShipPing] = inputRdd.resample.collect()
+    val outPings: DataFrame = inputDataFrame.resample(TEST_CONFIG)
 
     assertThatOutPingsMatchesExpectedPings(expectedPings, outPings)
   }
 
   @Test
   def whenResamplingPingsWithGapOver6HoursWhenSecondJourneyNotOnTimeBoundaryThenGapNotResampled()
```
Comment (oliverCausley) [inline spark/resample-ais/src/test/scala/ResamplerTest.scala]: Does this need an additional test for the other parameter (18hrs)?

Comment (JoelClemence) [reply spark/resample-ais/src/test/scala/ResamplerTest.scala]: Maybe

**Comment by JoelClemence** [comment]
@andyphelps Should this be closed?

**Pull request closed without merging**